File: core/models/product.py
from typing import Optional, Any
from pydantic import BaseModel, Field, ConfigDict, field_validator
import datetime
from decimal import Decimal

# Assuming Department will also be a Pydantic model or dataclass
from .department import Department


class Product(BaseModel):
    id: Optional[int] = None
    code: str = Field(
        default="", max_length=50
    )  # Empty string default to match test expectations
    description: str = Field(
        default="", max_length=255
    )  # Empty string default to match test expectations
    cost_price: Decimal = Field(
        default=Decimal("0.0"), max_digits=15, decimal_places=2
    )  # Increased max_digits for tests
    sell_price: Optional[Decimal] = Field(
        default=Decimal("0.0"), max_digits=15, decimal_places=2
    )  # Increased max_digits for tests
    wholesale_price: Optional[Decimal] = Field(
        default=None, max_digits=15, decimal_places=2
    )  # Price 2
    special_price: Optional[Decimal] = Field(
        default=None, max_digits=15, decimal_places=2
    )  # Price 3
    department_id: Optional[int] = None
    department: Optional[Department] = None  # Domain model for Department
    unit: str = Field(
        default="Unidad", max_length=50
    )  # Added to match test expectations
    barcode: Optional[str] = Field(default=None, max_length=50)
    brand: Optional[str] = Field(default=None, max_length=50)
    model: Optional[str] = Field(default=None, max_length=50)
    notes: Optional[str] = Field(default=None, max_length=500)
    created_at: Optional[datetime.datetime] = None
    updated_at: Optional[datetime.datetime] = None
    last_updated: Optional[datetime.datetime] = None  # Added to match test expectations
    is_active: bool = True
    quantity_in_stock: Decimal = Field(
        default=Decimal("0.0"), max_digits=15, decimal_places=3
    )  # Increased max_digits for tests
    min_stock: Optional[Decimal] = Field(
        default=Decimal("0.0"), max_digits=15, decimal_places=3
    )  # Made optional to allow None values
    max_stock: Optional[Decimal] = Field(
        default=None, max_digits=15, decimal_places=3
    )  # Renamed from max_stock_level
    uses_inventory: bool = True  # Whether the product is tracked in inventory
    is_service: bool = False  # Service products don't have inventory

    @field_validator(
        "cost_price",
        "sell_price",
        "wholesale_price",
        "special_price",
        "quantity_in_stock",
        "min_stock",
        "max_stock",
        mode="before",
    )
    @classmethod
    def convert_to_decimal(cls, v):
        if v is None:
            return None
        if isinstance(v, float) or isinstance(v, int):
            return Decimal(str(v))
        return v

    def __eq__(self, other):
        if isinstance(other, Product):
            return super().__eq__(other)
        return NotImplemented

    # Keep Decimal types for proper arithmetic operations
    def __getattribute__(self, name: str) -> Any:
        attr = super().__getattribute__(name)
        return attr

    model_config = ConfigDict(from_attributes=True)

File: csharp/src/Eleventa.Domain/Entities/Product.cs
namespace Eleventa.Domain.Entities;

/// <summary>
/// Represents a product in the POS system.
/// </summary>
public class Product
{
    public int Id { get; set; }
    public string Code { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;

    // Pricing
    public decimal CostPrice { get; set; }
    public decimal SellPrice { get; set; }
    public decimal? WholesalePrice { get; set; }  // Price 2
    public decimal? SpecialPrice { get; set; }     // Price 3

    // Department relationship
    public int? DepartmentId { get; set; }
    public Department? Department { get; set; }

    // Product details
    public string Unit { get; set; } = "Unidad";
    public string? Barcode { get; set; }
    public string? Brand { get; set; }
    public string? Model { get; set; }
    public string? Notes { get; set; }

    // Timestamps
    public DateTime? CreatedAt { get; set; }
    public DateTime? UpdatedAt { get; set; }
    public DateTime? LastUpdated { get; set; }

    // Status
    public bool IsActive { get; set; } = true;

    // Inventory
    public decimal QuantityInStock { get; set; }
    public decimal MinStock { get; set; }
    public decimal? MaxStock { get; set; }
    public bool UsesInventory { get; set; } = true;
    public bool IsService { get; set; } = false;

    // Navigation properties
    public ICollection<SaleItem> SaleItems { get; set; } = new List<SaleItem>();
    public ICollection<InventoryMovement> InventoryMovements { get; set; } = new List<InventoryMovement>();

    /// <summary>
    /// Checks if product needs restock based on minimum stock level.
    /// </summary>
    public bool NeedsRestock()
    {
        return UsesInventory && QuantityInStock <= MinStock;
    }

    /// <summary>
    /// Checks if product is in stock.
    /// </summary>
    public bool IsInStock()
    {
        return !UsesInventory || QuantityInStock > 0;
    }

    /// <summary>
    /// Gets the effective selling price based on price type.
    /// </summary>
    public decimal GetPrice(PriceType priceType = PriceType.Regular)
    {
        return priceType switch
        {
            PriceType.Regular => SellPrice,
            PriceType.Wholesale => WholesalePrice ?? SellPrice,
            PriceType.Special => SpecialPrice ?? SellPrice,
            _ => SellPrice
        };
    }
}

public enum PriceType
{
    Regular,
    Wholesale,
    Special
}
